### ISSUE TITLE
it's not possible to use InterleavedBufferAttribute if total stride in bytes is not divisible by the individual type byte size

### ISSUE THREAD

**Issue by autra** [mrdoob/three.js#14178]
##### Description of the problem

I receive from a server a buffer representing a pointclouds. Each individual points is composed from a position, and a variable set of other attributes (color, normals with different possible encoding, classification etc...), with variable types. My buffer topology is therefore

x1,y1,z1,r1,g1,b1,c1,x2,y2....,cn

I'd like to use `InterleavedBuffer` and `InterleavedBufferAttribute`.

They work well with [the examples on the website](https://threejs.org/examples/#webgl_buffergeometry_points_interleaved), because the total stride is 16 bytes, which is at the same time 16 Uint8  and 4 Float32 (hence the stride specified [there](https://github.com/mrdoob/three.js/blob/master/examples/webgl_buffergeometry_points_interleaved.html#L114) and [there](https://github.com/mrdoob/three.js/blob/master/examples/webgl_buffergeometry_points_interleaved.html#L115)).

But if I have a pointclouds composed of a position and a 1 byte classification, then I can create my classification buffer as such:

```js
var classificationTypedArray = new Uint8Array( arrayBuffer, 12);
var classificationIB = new THREE.InterleavedBuffer( classificationTypedArray, 13 /* (3* 32 + 8) / 8 */ );
var classificationBA = new THREE.InterleavedBufferAttribute( classificationIB, 1, 0);
``` 
but then I have a problem with my position buffer, I'd need to:
```js
var posTA = new Float32Array(arrayBuffer);
// This won't work of course.
var posIB = new THREE.InterleavedBuffer(posTA, 3.25 /* (3 * 32 + 8) / 32 */ );
var posBA = new THREE.InterleavedBufferAttribute( posIB, 3, 0);
```

I'd like to be able specify the stride in bytes instead of number of elements, like [vertexAttribPointer](https://developer.mozilla.org/en-US/docs/Web/API/WebGLRenderingContext/vertexAttribPointer).

working example (with a Float32 classification, so no problem)
 http://jsfiddle.net/zape7z10/7/

not working (what I want to do): 
http://jsfiddle.net/89m43qn9/1/

What would be the best way to achieve that? 

##### Three.js version

- [x] Dev
- [ ] r92
- [ ] ...

##### Browser

- [x] All of them
- [ ] Chrome
- [ ] Firefox
- [ ] Internet Explorer

##### OS

- [x] All of them
- [ ] Windows
- [ ] macOS
- [ ] Linux
- [ ] Android
- [ ] iOS

**Comment by Mugen87** [comment]
Your fiddle produces the following error:

> Uncaught RangeError: byte length of Float32Array should be a multiple of 4

I think this message is more or less self-explained. Besides, I've seen two errors in your first fiddle:

http://jsfiddle.net/zape7z10/7/

1. `InterleavedBuffer` has only two parameters. When creating `classifBf`, you assign a third one.
2. The third paramter of `InterleavedBufferAttribute` is the `offset` in the buffer where you attribute data starts. Both are `0` in your example which is wrong.

Both errors are fixed in this updated [fiddle](http://jsfiddle.net/zape7z10/8/). Now you also have a correct visual result.

> What would be the best way to achieve that?

Do it like in your first example. Or try to work with a padding to fill up the missing bytes.

**Comment by autra** [comment]
@Mugen87 Apparently I didn't express myself clearly. Let me try again:

Vanilla webgl takes a [stride](https://developer.mozilla.org/en-US/docs/Web/API/WebGLRenderingContext/vertexAttribPointer) that is a number of *bytes*, making the above use case possible (using a InterleavedBuffer with a 3 Float32 position InterleavedBufferAttribute and a one Int8 classification attributes). The stride would be therefore 13 (3*4 + 1).

In THREE.js, `InterleavedBuffer` takes a number of *elements* as `stride`, making this use case infeasible (yes I know I can pad my data as a workaround).

This is not a call for help (I should have left out the `what would the best way to achieve that?` above, I thought maybe I didn't see something). It's a feature request/bug report (not sure where to draw the line here): in my opinion three.js should mimic webgl api here.

What's your opinion here?

Thanks for your answer :-)

**Comment by WestLangley** [comment]
/ping @benaadams Can you please explain why you defined `stride`, `itemSize`, and `offset` with units of _elements_ rather than _bytes_? Thanks in advance.

**Comment by benaadams** [comment]
Non-aligned data causes issues as the driver then needs to rearrage the buffer to align the data before sending it to gpu.

> ### [Avoid Misaligned Vertex Data](https://developer.apple.com/library/archive/documentation/3DDrawing/Conceptual/OpenGLES_ProgrammingGuide/TechniquesforWorkingwithVertexData/TechniquesforWorkingwithVertexData.html#//apple_ref/doc/uid/TP40008793-CH107-SW7)
>
> When you are designing your vertex structure, align the beginning of each attribute to an offset that is either a multiple of its component size or 4 bytes, whichever is larger. When an attribute is misaligned, iOS must perform additional processing before passing the data to the graphics hardware.
>
> In Figure 8-4, the position and normal data are each defined as three short integers, for a total of six bytes. The normal data begins at offset 6, which is a multiple of the native size (2 bytes), but is not a multiple of 4 bytes. If this vertex data were submitted to iOS, iOS would have to take additional time to copy and align the data before passing it to the hardware. To fix this, explicitly add two bytes of padding after each attribute.

![image](https://user-images.githubusercontent.com/1142958/43858232-6d41ef64-9b45-11e8-9c59-fbea5deb490d.png)
